### PR TITLE
Adding a warning header when a license is about to expire

### LINKED ISSUE #60562: Provide warning headers when using a licensed feature and the license expiration is looming
When a user is using a licensed feature and license expiration is pending, we could provide a warning header informing them that they are using such a feature that they are about to lose access to.

Relates elastic/kibana#74009.

### PR DESCRIPTION
**Pull request by BigPandaToo**
This implementation will add the warning header 
if the license is going to expire in less than 
{`LICENSE_EXPIRATION_WARNING_PERIOD`} days. 
The messages added:

`Warning: 299 Elasticsearch-8.0.0-###"Your license will expire in [N] days. Contact your administrator or update your license for continued use of features"`

or

`Warning: 299 Elasticsearch-8.0.0-### "Your license expires today. Contact your administrator or update your license for continued use of features"`

If license has expired less than 
{`GRACE_PERIOD_DURATION`} days ago following
 warning is added:

`Warning: 299 Elasticsearch-8.0.0-### "Your license expired on ["EEEE, MMMM dd, yyyy" ]. Contact your administrator or update your license for continued use of features"`

Both {`LICENSE_EXPIRATION_WARNING_PERIOD`} 
and {`GRACE_PERIOD_DURATION`} are currently 7 days.

The message will be added to each request unless 
authentication fails.
Note: with this change all warning headers will be removed 
from a response if authentication fails.

Resolves #60562

### CODE REVIEW

**Comment by BigPandaToo** [comment]
@elasticmachine update branch

**Comment by BigPandaToo** [comment]
@elasticmachine update branch

**Comment by BigPandaToo** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Comment by BigPandaToo** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Comment by BigPandaToo** [comment]
@elasticmachine update branch

**Comment by BigPandaToo** [comment]
@elasticmachine update branch

**Comment by BigPandaToo** [comment]
@elasticmachine update branch

File: x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java
Comment (jkakavas) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java:407]: s/avILble/available. Nit: also change the description of the Status class above to add the expiration date

File: x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java
Comment (jkakavas) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java:526]: I think we'd need to access this by calling `checkAgainstStatus` instead of accessing the licenseExpiryDate  field directly

File: x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java
Comment (jkakavas) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java:525]: nit: `if (feature....`

File: x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java
Comment (jkakavas) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java:528]: I'm not sure about the wording here but I'll defer to a native speaker and maybe get Fabio's input too? 

Maybe s/update/renew ?  and s/for continued use of/in order to continue using the licensed ?

**Review by jkakavas — COMMENTED** [review]
I left a few comments. I also think we need to elaborate on the PR description ( it's not clear what `through SECURITY_IP_FILTER`  means ) and add a couple of sentences on why we add the check on `checkFeature` and more especially why we think this is enough for solving the issue this PR is meant to. 

We also need some tests that ensure we are indeed adding the warning header under the circumstances we expect it to be added

File: x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java
Comment (BigPandaToo) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java:528]: This is the wording Kibana has. I talked to Fabio and he suggested us to be consistant with wording when possible.

Comment (jkakavas) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java:528]: SGTM then!

**Comment by BigPandaToo** [comment]
@elasticmachine update branch

File: x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java
Comment (tvernum) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java:526]: If we're in the grace period, shouldn't this say `has expired` ?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java
Comment (tvernum) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java:648]: Why is this dependent on the `GRACE_PERIOD` ? It seems to be a co-incidence that they're both 7days, but they shouldn't be tied to one another.

**Review by tvernum — CHANGES_REQUESTED** [review]
It doesn't look like there are any tests for the new warning.

I think we need some (at the very least to verify that the time calculations are handled correctly).

File: x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java
Comment (BigPandaToo) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java:526]: I'll double-check with Fabio. This is the only message Kibana has and the idea was to be consistent when possible.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java
Comment (BigPandaToo) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java:648]: Good point, I'll add new const

File: x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java
Comment (bytebilly) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java:526]: I don't think we have a grace period anymore. AFAIK we have warnings before the expiration date, and no service after that. Could you please double check what `isLicenseExpiring()` assumes?

Comment (tvernum) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java:526]: We still have a grace period. We may not in the future, but we do now.

Comment (bytebilly) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java:526]: _(sorry I cannot find the description of the full approach, so it may be already this)_

What about having two separate checks with two separate messages?

- 7 days before expiration — expiration: "will expire in N days"
- expiration — end of grace period: "has expired N days ago"

Comment (BigPandaToo) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java]: I changed the behaviour to have 3 different messages:
 - 7 days before expiration — expiration: "will expire in N days"
- <1 days before expiration: "expires today"
- > expiration: "has expired N days ago"
If grace period is over this code should be irrelevant as the license is not active.

Comment (BigPandaToo) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java]: Also, addad a test

**Comment by tvernum** [comment]
It looks like we return the warning header even if authentication fails. I don't think we should do that.

**Review by bytebilly — COMMENTED** [review]
@BigPandaToo sorry I'm not confident I can get the exact behavior of this feature looking at the code only.
Could you please briefly describe high-level actions?
I'd want to get answers to questions like:
- when the message is sent (time based)
- where the message is sent (every request, requests for paid features only, etc)
- what is sent (message content, code, etc)

That would be very helpful to provide some feedback.
Thanks!

File: x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java
Comment (jkakavas) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java:660]: Given what this does, would it make sense to name it `shouldWarnAboutLicense()` ?

File: x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/license/LicensingTests.java
Comment (jkakavas) [inline x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/license/LicensingTests.java:306]: Do we have to do the double update dance here? `setLicensingExpirationDate` is the only thing that calls update with anything other than Long.MAX_VALUE, what cluster state could trigger the license state (Expiration date) to change ?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java
Comment (jkakavas) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java:530]:
```suggestion
            final String expiryMessage = days == 0 ? "expires today" :
```

File: x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java
Comment (jkakavas) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java:531]:
```suggestion
                (days > 0 ? String.format(Locale.ROOT, "will expire in [%d] days", days):
```

**Review by jkakavas — COMMENTED** [review]
We still need a proper PR description to accompany the commit once this is merged.

**Comment by BigPandaToo** [comment]
@elasticmachine update branch

**Comment by BigPandaToo** [comment]
@elasticmachine update branch

File: x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/license/LicensingTests.java
Comment (BigPandaToo) [reply x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/license/LicensingTests.java:306]: Probably, not. Removed the second update, left hte assertBusy though.

**Comment by BigPandaToo** [comment]
> It looks like we return the warning header even if authentication fails. I don't think we should do that.

Changing the approach. Moving the warning adding block to SecurityRestFilter. There are pros and cons though:
pros:
 - we can avoid adding the headers when authentication fails
 - the message is sent with all requests when security is enabled
cons:
 - no message when security is not enabled

**Comment by BigPandaToo** [comment]
> @BigPandaToo sorry I'm not confident I can get the exact behavior of this feature looking at the code only.
> Could you please briefly describe high-level actions?
> I'd want to get answers to questions like:
> 
> * when the message is sent (time based)
> * where the message is sent (every request, requests for paid features only, etc)
> * what is sent (message content, code, etc)
> 
> That would be very helpful to provide some feedback.
> Thanks!

@bytebilly 
I have changed the code as the previous approach didn't allow avoiding messaging if authentication failed. Please, see the new description.

**Review by bytebilly — APPROVED** [review]
The high-level approach LGTM 👍

**Comment by tvernum** [comment]
> no message when security is not enabled

That sounds like a pretty big downside.
I think we can do something like this -> https://github.com/tvernum/elasticsearch/commit/d8f36cf52f8 <- and then stick with the original version that relied on `XPackLicenseState`.

It would have the effect of removing all warning headers, but I think that's actually preferred, (see also: https://github.com/elastic/elasticsearch/pull/55774)

@tbrooks8, since you seem to be the primary maintainer of `DefaultRestChannel` do you think the commit I linked to just above is an appropriate way to prevent `Warning` headers from being sent on authentication failure responses.

**Comment by BigPandaToo** [comment]
@elasticmachine update branch

File: x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java
Comment (droberts195) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java:516]: Not necessary?  This method doesn't appear to use `Math.abs()`.

**Comment by BigPandaToo** [comment]
@elasticmachine update branch

File: x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java
Comment (BigPandaToo) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java:516]: Leftover from the previous iteration, but will come back handy with the next one )

**Comment by BigPandaToo** [comment]
@elasticmachine update branch

File: x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java
Comment (tvernum) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java:536]: For expired licenses, I think it's preferable to just print out the date that is expired, rather than calculate how lonh ago that was.
That's what the log warning does and it seems neater for this case.

> LICENSE [EXPIRED] ON [SUNDAY, NOVEMBER 29, 2020]

The number of days makes sense for "will expire" because it creates a sense of urgency, but once it's expired I think the date is a better option.

Comment (tvernum) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java:536]: I think we should be doing the first round of calculations in millis, not in days otherwise we potentially says "expires today" _after_ the expiry date (because it's less than 1 day expired)

```suggestion
    
        if (feature.minimumOperationMode.compareTo(OperationMode.BASIC) > 0) {
            final long expiryMillis = licenseExpirationDate - now;        
            if (expiryMillis < LICENSE_EXPIRATION_WARNING_PERIOD.getMillis()) {
                final String expiryMessage;
                if (expiryMillis <= 0) {
                    expiryMessage = "expired on [" + DATE_FORMATTER.formatMillis(expirationMillis) + "]";
                } else {
                    final long days = TimeUnit.MILLISECONDS.toDays(expiryMillis);
                    expiryMessage = days == 0 ? "expires today": 
                       String.format(Locale.ROOT, "will expire in [%d] days", days);
                }
```

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/SecurityRestFilter.java
Comment (tvernum) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/SecurityRestFilter.java:112]: Do we have a test for this anywhere?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java
Comment (BigPandaToo) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java]: Changed the wording to "license expired on [EEEE, MMMM DD, YYYY]". For the sace of keeping the headers relatively short kept the rest of the message the same.

Comment (BigPandaToo) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java]: Done

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/SecurityRestFilter.java
Comment (BigPandaToo) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/SecurityRestFilter.java:112]: Added a test to verify Warnings are removed if authN fails.

**Review by tvernum — APPROVED** [review]
LGTM, but I'd like to hear from @tbrooks8 before we merge.
(and if we can switch to `assertThat`, that would be good too)

**Comment by BigPandaToo** [comment]
@elasticmachine update branch

**Review by Tim-Brooks — APPROVED** [review]
LGTM

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/SecurityRestFilter.java
Comment (jkakavas) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/SecurityRestFilter.java:107]: Do we want to remove _all_ warning headers on authentication failure?

File: x-pack/plugin/core/src/test/java/org/elasticsearch/test/SecuritySettingsSourceField.java
Comment (jkakavas) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/test/SecuritySettingsSourceField.java:13]: I think you can use a randomAlphaNumericOfLength , or a permutation of `TEST_PASSWORD` plus some random string for whenever we need this, other than defining it here. That's just personal preference, no strong views, feel free to keep as is if you prefer it

File: x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/license/LicensingTests.java
Comment (jkakavas) [inline x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/license/LicensingTests.java:216]: use assertThat ?

File: x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/license/LicensingTests.java
Comment (jkakavas) [inline x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/license/LicensingTests.java:211]: nit: Do we need all the empty lines?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java
Comment (jkakavas) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java:535]: I Guess we are fine with the message in warnigns and subtleties of `toDays` and being close to midnight in a given timezone , right? I think we are, just asking for completeness.

File: x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/license/LicensingTests.java
Comment (jkakavas) [inline x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/license/LicensingTests.java:278]: use assertThat ?

File: x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/license/LicensingTests.java
Comment (jkakavas) [inline x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/license/LicensingTests.java:258]: nit suggestion : testNoWarningHeaderWhenAuthenticationFailed

**Review by jkakavas — COMMENTED** [review]
Just some nits suggestions and a question regarding if removing _all_ warning headers on authentication failure is intentional

**Comment by BigPandaToo** [comment]
> Just some nits suggestions and a question regarding if removing _all_ warning headers on authentication failure is intentional

@jkakavas Yes, it is intentional (see Tim's comment above). If Authentication fails it does make sense to not reveal any additional information

File: x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/license/LicensingTests.java
Comment (BigPandaToo) [reply x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/license/LicensingTests.java:211]: Yes, it is intentional (see Tim's comment above). If Authentication fails it does make sense to not reveal any additional information

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/SecurityRestFilter.java
Comment (BigPandaToo) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/SecurityRestFilter.java:107]: Yes, it is intentional (see Tim's comment above). If Authentication fails it does make sense to not reveal any additional information

File: x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java
Comment (BigPandaToo) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java:535]: I believe we are :)

**Review by jkakavas — APPROVED** [review]
LGTM

**Comment by BigPandaToo** [comment]
@elasticmachine update branch

**Pull request merged**
